packages/babel-helper-create-class-features-plugin/src/fields.js

```diff
@@ -85,24 +85,36 @@ const privateNameVisitor = {
     const { privateNamesMap } = this;
     const body = path.get("body.body");
 
+    const visiblePrivateNames = new Map(privateNamesMap);
     for (const prop of body) {
-      if (!prop.isPrivate()) {
-        continue;
-      }
-      if (!privateNamesMap.has(prop.node.key.id.name)) continue;
+      if (!prop.isPrivate()) continue;
+      visiblePrivateNames.delete(prop.node.key.id.name);
+    }
 
-      // This class redeclares the private name.
-      // So, we can only evaluate the things in the outer scope.
-      path.traverse(privateNameInnerVisitor, this);
-      path.skip();
-      break;
+    // If the class doesn't redeclare any private fields, we can continue with
+    // our overall traversal.
+    if (visiblePrivateNames.size === privateNamesMap.size) {
+      return;
     }
+
+    // This class redeclares some private field. We need to process the outer
+    // environment with access to all the outer privates, then we can process
+    // the inner environment with only the still-visible outer privates.
+    path.traverse(privateNameNestedVisitor, this);
```

Comment:
@JLHwung Maybe it would be enough to remove this line to fix it?

Reply:
I don't think so but I haven't tried. The nested private name is still allowed as long as it does not accessing outer class private identifiers. I think we shall whitelist computed element names for proper private name resolving behaviour.

```js
class Foo {
  #foo = 1;

  test() {
    class Nested {
      #foo = 2;
      bar = this.#foo;
    }

    new Nested;
  }
}
(new Foo).test();
```